ngraph/core/src/op/gather.cpp

```diff
@@ -140,6 +140,163 @@ shared_ptr<Node> op::v1::Gather::clone_with_new_inputs(const OutputVector& new_a
+    auto axis_rank = axis_pshape.rank();
+
+    if (axis_rank.is_static() && axis_pshape.is_static())
+    {
+        const auto axis_is_scalar = axis_rank.get_length() == 0;
+        const auto axis_has_one_elem =
+            axis_rank.get_length() == 1 && axis_pshape[0].get_length() == 1;
+        NODE_VALIDATION_CHECK(this,
+                              axis_is_scalar || axis_has_one_elem,
+                              "Axes input must be scalar or have 1 element (shape: ",
+                              axis_pshape,
+                              ").");
+    }
+
+    int64_t axis = get_axis();
+    int64_t batch_dims = get_batch_dims();
+    NODE_VALIDATION_CHECK(this, axis != AXIS_NOT_SET_VALUE, "The axis is not specified");
+    NODE_VALIDATION_CHECK(this,
+                          axis >= 0 && axis < data_rank.get_length(),
+                          "The axis must be => 0 and < data_rank. But instead got axis = ",
+                          axis,
+                          " data_rank = ",
+                          data_rank.get_length());
+
+    if (data_rank.is_static() && axis != AXIS_NOT_SET_VALUE)
+    {
+        NODE_VALIDATION_CHECK(this,
+                              axis >= batch_dims,
+                              "The axis must be => batch_dims. But instead got: axis = ",
+                              axis,
+                              ", batch_dims = ",
+                              batch_dims);
+    }
+
+    if (data_pshape.rank().is_static() && indices_pshape.rank().is_static())
+    {
+        auto out_rank =
+            data_pshape.rank().get_length() + indices_pshape.rank().get_length() - 1 - batch_dims;
+        std::vector<Dimension> result_dims(out_rank);
+        PartialShape output_pshape(result_dims);
+
+        // implementation of out_shape formula
+        // data.shape[:batch_dims] + data.shape[batch_dims:axis] + indices.shape[batch_dims:] +
+        // data.shape[axis + 1:]
+        int i = 0;
+        for (; i < batch_dims; i++)
+        {
+            NODE_VALIDATION_CHECK(this,
+                                  data_pshape[i].compatible(indices_pshape[i]),
+                                  "Shapes ",
+                                  data_pshape,
+                                  " and ",
+                                  indices_pshape,
+                                  " are not consistent. data and indices must have equal or "
+                                  "intersecting sizes until batch_dims");
+
+            output_pshape[i] = data_pshape[i] & indices_pshape[i];
+        }
+        for (; i < axis; i++)
+        {
+            output_pshape[i] = data_pshape[i];
+        }
+        for (; i < axis + indices_rank.get_length() - batch_dims; i++)
+        {
+            output_pshape[i] = indices_pshape[batch_dims - axis + i];
+        }
+        for (; i < out_rank; i++)
+        {
+            output_pshape[i] = data_pshape[batch_dims + 1 - indices_rank.get_length() + i];
```

Comment:
What if `batch_dims + 1 - indices_rank.get_length() + i` is bigger than rank of `data_pshape`?

Reply:
that's a good question, in that case we will have out of bound error. 

But if `batch_dims` is < than indices_rank and data_rank then we will newer exceed out of bound. I will add check for `batch_dims`

Reply:
if `batch_dims < data_rank` and `batch_dims < indices_rank` then `i` and `batch_dims + 1 - indices_rank.get_length() + i` is within acceptable range. 

E.g. max value of `batch_dims + 1 - indices_rank.get_length() + i` is `data_rank_rank - 1- batch_dims - axis`. If values are verified previously no need for additional check here.